cns/restserver/ipam.go

```diff
@@ -478,101 +627,230 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+			}
+		}
+		//nolint:goerr113 // return error
+		return podIPInfo, fmt.Errorf("not all requested ips %v were found/available in the pool", desiredIPAddresses)
+	}
+
+	logger.Printf("[AssignDesiredIPConfigs] Successfully assigned all desired IPs for pod %+v", podInfo)
+	return podIPInfo, nil
 }
 
-func (service *HTTPRestService) AssignAnyAvailableIPConfig(podInfo cns.PodInfo) (cns.PodIpInfo, error) {
+// Assigns an available IP from each NC on the NNC. If there is one NC then we expect to only have one IP return
+// In the case of dualstack we would expect to have one IPv6 from one NC and one IPv4 from a second NC
+func (service *HTTPRestService) AssignAvailableIPConfigs(podInfo cns.PodInfo) ([]cns.PodIpInfo, error) {
 	service.Lock()
 	defer service.Unlock()
-
+	// Sets the number of IPs needed equal to the number of NCs so that we can get one IP per NC
+	numOfIPsNeeded := len(service.state.ContainerStatus)
+	// Creates a slice of PodIpInfo with the size as number of NCs to hold the result for assigned IP configs
+	podIPInfo := make([]cns.PodIpInfo, numOfIPsNeeded)
+	// This map is used to store whether or not we have found an available IP from an NC when looping through the pool
+	availableIPs := make(map[string]cns.IPConfigurationStatus)
+
+	// Searches for available IPs in the pool
 	for _, ipState := range service.PodIPConfigState {
-		if ipState.GetState() == types.Available {
-			if err := service.assignIPConfig(ipState, podInfo); err != nil {
-				return cns.PodIpInfo{}, err
-			}
+		_, found := availableIPs[ipState.NCID]
+		// Checks if we haven't already found an IP from that NC and checks if the current IP is available
+		if found || ipState.GetState() != types.Available {
+			continue
+		}
+		availableIPs[ipState.NCID] = ipState
+		// Once one IP per container is found break out of the loop and stop searching
+		if len(availableIPs) == numOfIPsNeeded {
+			break
+		}
+	}
 
-			podIPInfo := cns.PodIpInfo{}
-			if err := service.populateIPConfigInfoUntransacted(ipState, &podIPInfo); err != nil {
-				return cns.PodIpInfo{}, err
-			}
+	// Checks to make sure we found one IP for each NC
+	if len(availableIPs) != numOfIPsNeeded {
+		//nolint:goerr113 // return error
+		return podIPInfo, fmt.Errorf("not able to find one available IP per NC in IP pool")
+	}
 
+	failedToAssignIP := false
+	numIPConfigsAssigned := 0
+	// assigns all IPs in the map to the pod
+	for id := range availableIPs {
+		if err := service.assignIPConfig(availableIPs[id], podInfo); err != nil {
+			logger.Errorf(err.Error())
+			failedToAssignIP = true
+			break
+		}
+
+		if err := service.populateIPConfigInfoUntransacted(availableIPs[id], &podIPInfo[numIPConfigsAssigned]); err != nil {
+			logger.Errorf(err.Error())
+			failedToAssignIP = true
+			break
+		}
+		numIPConfigsAssigned++
+		// Checks if we no longer have any IPs to be added and returns if we don't
+		if numIPConfigsAssigned == numOfIPsNeeded {
```

Comment:
This isn't for optimizing, more for finding the happy path. Right now if we never pass this if statement it will go and unassign everything since we didn't assign the number of IPs needed.

Reply:
See the other comments for this func. You would know why we don't need this.